core/types/pow/ipow.go

```diff
@@ -0,0 +1,171 @@
+package pow
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"github.com/Qitmeer/qitmeer-lib/common/hash"
+	"math/big"
+)
+
+// proof data length 188
+const POW_LENGTH = 188
+//except pow type 4bytes and nonce 8 bytes 176 bytes
+const PROOFDATA_LENGTH = 176
+type PowType int
+type PowBytes [POW_LENGTH]byte
+
+const (
+	//pow type enum
+	BLAKE2BD PowType = 0
+	CUCKAROO PowType = 1
+	CUCKATOO PowType = 2
+)
+
+var PowMapString = map[PowType]interface{}{
+	BLAKE2BD : "blake2bd",
+	CUCKAROO : "cuckaroo",
+	CUCKATOO : "cuckatoo",
+}
+
+type PowResult struct {
+    Nonce     uint64     `json:"nonce"`
+    PowType   string     `json:"pow_type"`
+    ProofData string     `json:"proof_data"`
+}
+
+type ProofDataType [PROOFDATA_LENGTH]byte
+
+func (this *ProofDataType) String() string{
+	return hex.EncodeToString(this[:])
+}
+
+func (this *ProofDataType) Bytes() []byte{
+	return this[:]
+}
+
+type PowConfig struct {
+	// PowLimit defines the highest allowed proof of work value for a block
+	// as a uint256.
+	Blake2bdPowLimit *big.Int
+	// PowLimitBits defines the highest allowed proof of work value for a
+	// block in compact form.
+	// highest value is mean min difficulty
+	Blake2bdPowLimitBits uint32
+
+	// cuckoo difficulty calc params  min difficulty
+	CuckarooMinDifficulty uint32
+	CuckatooMinDifficulty uint32
+	// solotion difficulty diff = (scale * 1<<64) u128 / cuckoohash as u128
+	CuckarooDiffScale uint64
+	CuckatooDiffScale uint64
+
+	//percent of every pow sum of them must be 100
+	CuckarooPercent int
+	CuckatooPercent int
+	Blake2bDPercent int
+}
+
+type IPow interface {
+	// verify result difficulty
+	Verify(headerWithoutProofData []byte,targetDiff uint64,powConfig *PowConfig) error
+	//set header nonce
+	SetNonce(nonce uint64)
+	//calc next diff
+	GetNextDiffBig(weightedSumDiv *big.Int,oldDiffBig *big.Int,currentPowPercent *big.Int,param *PowConfig) *big.Int
+	GetNonce() uint64
+	GetPowType() PowType
+	//set pow type
+	SetPowType(powType PowType)
+	GetProofData() string
+	//set proof data
+	SetProofData([]byte)
+	GetBlockHash(data []byte) hash.Hash
```

Comment:
I don't think the IPOW interface should contain a different way to calculate the `blockhash`.  the ProofOfPow data can differ from the different POW type. but the *Block*hash should always is hash(BlockheaderWithoutProof+ProofData).